dask_expr/_collection.py

```diff
@@ -828,6 +829,9 @@ def set_index(
         elif other == self.index.name:
             return self
 
+        if not sort:
+            return new_collection(SetIndexBlockwise(self.expr, other, drop, None))
+
```

Comment:
I think I missed this detail from #230, but don't we need to satisfy `npartitions` when `sort=False` and when `sorted=True`?

Reply:
I think the idea behind that was to have a fastpath. I’d rather raise if sort is false and partitions is given instead of doing a repartition

Reply:
There is no performance benefit over doing an explicit repartition

Reply:
Yeah, raising an error is fine with me, I just don't want to silently ignore the `npartitions` argument. Does that make sense?

Reply:
Yep, added a check